hal_nxp: Update to SDK2.8

mmahadevan108 opened the pull request.

mmahadevan108 committed:
soc: nxp_imx: Update i.MX8M due to update to SDK drivers

Some of the clock macros have changed with update of SDK drivers
to SDK 2.8

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
west: nxp: Update to SDK 2.8

MCUX SDK drivers updates to revision 2.8

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented:
cc @MaureenHelm

henrikbrixandersen commented on west.yml:
```diff
@@ -98,7 +98,7 @@ manifest:
       path: tools/net-tools
     - name: hal_nxp
-      revision: e13b23c06850530c6c34208905d58e7a6a12cdd5
+      revision: zephyrproject-rtos/hal_nxp/pull/56
```
```suggestion
      revision: pull/56/head
```

mmahadevan108 replied:
Done

henrikbrixandersen commented:
Looks like at least `drivers/ethernet/eth_mcux.c` needs updating to use the new MCUX SDK.

mmahadevan108 commented:
> Looks like at least `drivers/ethernet/eth_mcux.c` needs updating to use the new MCUX SDK.

We will have to look into this as a separate PR.

MaureenHelm commented:
> > Looks like at least `drivers/ethernet/eth_mcux.c` needs updating to use the new MCUX SDK.
> 
> We will have to look into this as a separate PR.

It needs to be fixed in this PR

mmahadevan108 committed:
drivers: eth_mcux: Update to match changes to the SDK driver

1. Merged the single/multiple ring(s) APIs, now these APIs
can handle both which requires passing a ring id.
2. Updated timestamp handling to match the new timestamp
implementation in the SDK driver.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

jukkar commented on drivers/ethernet/eth_mcux.c:
```diff
@@ -663,51 +663,6 @@ static bool eth_get_ptp_data(struct net_if *iface, struct net_pkt *pkt,
-		 * and the actual gPTP header is in the second net_buf.
-		 * In RX, the Ethernet header + other headers are in the
-		 * first net_buf.
-		 */
-		if (is_tx) {
-			if (pkt->frags->frags == NULL) {
-				return false;
-			}
-
-			hdr = (struct gptp_hdr *)pkt->frags->frags->data;
-		} else {
-			hdr = (struct gptp_hdr *)(pkt->frags->data +
-							   eth_hlen);
-		}
-
-		ptpTsData->version = hdr->ptp_version;
-		memcpy(ptpTsData->sourcePortId, &hdr->port_id,
-		       kENET_PtpSrcPortIdLen);
-		ptpTsData->messageType = hdr->message_type;
-		ptpTsData->sequenceId = ntohs(hdr->sequence_id);
-
-#ifdef CONFIG_ETH_MCUX_PHY_EXTRA_DEBUG
-		LOG_DBG("PTP packet: ver %d type %d len %d seq %d",
-			ptpTsData->version,
-			ptpTsData->messageType,
-			ntohs(hdr->message_length),
-			ptpTsData->sequenceId);
-		LOG_DBG("  clk %02x%02x%02x%02x%02x%02x%02x%02x port %d",
-			hdr->port_id.clk_id[0],
-			hdr->port_id.clk_id[1],
-			hdr->port_id.clk_id[2],
-			hdr->port_id.clk_id[3],
-			hdr->port_id.clk_id[4],
-			hdr->port_id.clk_id[5],
-			hdr->port_id.clk_id[6],
-			hdr->port_id.clk_id[7],
-			ntohs(hdr->port_id.port_number));
-#endif
-	}
-
```
You are removing lot of code here. Is gPTP and timestamping support really tested properly after this change?
I am asking because `frdm-k64f` has been one of the reference devices that support time sync in TSN and it would be important that gPTP continues to work properly in the future with this hw.

mmahadevan108 replied:
I am not clear how to run the gPTP tests. Can you give me some guidance. I tried following the instructions provided, however I see an error `Group ptp not found` when running `sudo ./gptp` command

jukkar replied:
One option for testing is to have Linux <-> Zephyr gPTP connectivity and for that I suggest linuxptp from http://linuxptp.sourceforge.net/. Linuxptp supports software timestamping so it can be run in normal PC.

The error `Group ptp not found` sounds like the ptp group is missing and your account is not part of that group.

The other option is to have two Zephyr devices that support gPTP and connect them together. Direct connection might work too, but if not then you need to connect them both to a Ethernet switch.

mmahadevan108 replied:
Thanks, I have installed `linuxptp`. Can you provide some guidance on how to use it for testing frdm-k64F. I have the `samples/net/gptp` application running on the frdm-k64F board

mmahadevan108 replied:
I tried with 2 boards running `samples/net/gptp`. I see the below printed on both boards and nothing more.
```
[00:00:00.269,000] <wrn> net_if: iface 0x80000f60 is down
[00:00:00.273,000] <inf> net_config: Initializing network
[00:00:00.273,000] <inf> net_config: Waiting interface 1 (0x80000f60) to be up...
[00:00:01.271,000] <wrn> net_gptp: Reset Pdelay requests
[00:00:01.273,000] <wrn> net_if: iface 0x80000f60 is down
uart:~$
```

jukkar replied:
You can use `net gptp` and `net gptp 1` net-shell commands to see some information about gPTP status.

mmahadevan108 replied:
I  am unable to type anything in the serial terminal window

hakehuang replied:
in my testing there are some problem
![image](https://user-images.githubusercontent.com/800198/96839731-28ec3100-147c-11eb-8784-35ca8de13988.png)

I run the test script in windows
.gptp <mac address> -F gptp_cfg.ini

@jukkar  can you have a check on this? looks like the gptp application is not work well in ptpv2 IEEE1588

jukkar replied:
> @jukkar can you have a check on this? looks like the gptp application is not work well in ptpv2 IEEE1588

My frdm-k64f board just refuses the take the Ethernet link up so I am seeing the same issue as @mmahadevan108 few comments above.

Instead I used Atmel sam-e70-xplained board against Avnu gPTP daemon running in Linux, and everything is working fine there. Zephyr becomes slave and linux host became master, time sync is ok.
Can you do your tests without the lines 666-710 removed (I commented these above)? The removal of those lines looks suspicious and I would like to get confirmation whether they break gPTP or not.

hakehuang replied:
comment out those lines does not help, I still get the same result.  and the board console output is

```
[00:06:55.751,000] <wrn> net_gptp: Reset Pdelay requests
[00:06:56.756,000] <wrn> net_gptp: Reset Pdelay requests
[00:06:57.760,000] <wrn> net_gptp: Reset Pdelay requests
[00:06:58.764,000] <wrn> net_gptp: Reset Pdelay requests
[00:06:59.769,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:00.773,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:01.777,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:02.782,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:03.786,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:04.790,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:05.794,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:06.799,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:07.803,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:08.807,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:09.812,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:10.816,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:11.820,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:12.825,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:13.829,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:14.833,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:15.837,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:16.842,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:17.846,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:18.850,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:19.855,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:20.859,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:21.863,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:22.868,000] <wrn> net_gptp: Reset Pdelay requests
[00:07:23.872,000] <wrn> net_gptp: Reset Pdelay requests

```

hakehuang replied:
board networking config

```
[00:00:00.003,000] <wrn> net_if: iface 0x20000f40 is down
[00:00:00.007,000] <inf> net_config: Initializing network
[00:00:00.007,000] <inf> net_config: Waiting interface 1 (0x20000f40) to be up...
[00:00:01.005,000] <wrn> net_gptp: Reset Pdelay requests
[00:00:01.008,000] <wrn> net_if: iface 0x20000f40 is down
[00:00:02.010,000] <wrn> net_gptp: Reset Pdelay requests
[00:00:02.012,000] <wrn> net_if: iface 0x20000f40 is down
[00:00:03.001,000] <inf> net_config: Interface 1 (0x20000f40) coming up
[00:00:03.001,000] <inf> eth_mcux: ETH_0 enabled 100M full-duplex mode.
[00:00:03.001,000] <inf> net_config: IPv4 address: 192.168.0.103
[00:00:03.014,000] <wrn> net_gptp: Reset Pdelay requests
[00:00:03.102,000] <inf> net_config: IPv6 address: 2001:db8::1
uart:~$ net gptp 1
Port id    : 1 (DISABLED)
Interface  : 0x20000f40 [1]
Clock id   : 02:04:9f:ff:fe:31:96:09
Version    : 2
AS capable : no

Configuration:
Time synchronization and Best Master Selection enabled        : yes
The port is measuring the path delay                          : no
One way propagation time on the link attached to this port    : 0 ns
Propagation time threshold for the link attached to this port : 100000 ns
Estimate of the ratio of the frequency with the peer          : 1
Asymmetry on the link relative to the grand master time base  : %ld
Maximum interval between sync messages                        : %lu
Maximum number of Path Delay Requests without a response      : 3
Current Sync sequence id for this port                        : 24570
Current Path Delay Request sequence id for this port          : 50891
Current Announce sequence id for this port                    : 15514
Current Signaling sequence id for this port                   : 22146
Whether neighborRateRatio needs to be computed for this port  : yes
Whether neighborPropDelay needs to be computed for this port  : yes
Initial Announce Interval as a Logarithm to base 2            : 0
Current Announce Interval as a Logarithm to base 2            : 0
Initial Sync Interval as a Logarithm to base 2                : -4
Current Sync Interval as a Logarithm to base 2                : -4
Initial Path Delay Request Interval as a Logarithm to base 2  : 0
Current Path Delay Request Interval as a Logarithm to base 2  : 0
Time without receiving announce messages before running BMCA  : 1 ms (3)
Time without receiving sync messages before running BMCA      : %lu ms (1)
Sync event transmission interval for the port                 : %lu ms
Path Delay Request transmission interval for the port         : %lu ms
BMCA default priority1                                        : 248
BMCA default priority2                                        : 248

Runtime status:
Current global port state                                : DISABLED
Path Delay Request state machine variables:
Current state                                    : WAIT_RESP
Initial Path Delay Response Peer Timestamp       : %lu
Initial Path Delay Response Ingress Timestamp    : %lu
Path Delay Response messages received            : 0
Path Delay Follow Up messages received           : 0
Number of lost Path Delay Responses              : 3
Timer expired send a new Path Delay Request      : 0
NeighborRateRatio has been computed successfully : 0
Path Delay has already been computed after init  : 1
Count consecutive reqs with multiple responses   : 0
Path Delay Response state machine variables:
Current state                                    : INITIAL_WAIT_REQ
SyncReceive state machine variables:
Current state                                    : DISCARD
A Sync Message has been received                 : no
A Follow Up Message has been received            : no
A Follow Up Message timeout                      : no
Time at which a Sync Message without Follow Up
                             will be discarded   : %lu
SyncSend state machine variables:
Current state                                    : INITIALIZING
A MDSyncSend structure has been received         : no
The timestamp for the sync msg has been received : no
PortSyncSyncReceive state machine variables:
Current state                                    : DISCARD
Grand Master / Local Clock frequency ratio       : 0.000000
A MDSyncReceive struct is ready to be processed  : no
Expiry of SyncReceiptTimeoutTimer                : no
PortSyncSyncSend state machine variables:
Current state                                    : TRANSMIT_INIT
Follow Up Correction Field of last recv PSS      : %ld
Upstream Tx Time of the last recv PortSyncSync   : %lu
Rate Ratio of the last received PortSyncSync     : 0.000000
GM Freq Change of the last received PortSyncSync : 0.000000
GM Time Base Indicator of last recv PortSyncSync : 0
Received Port Number of last recv PortSyncSync   : 0
PortSyncSync structure is ready to be processed  : no
Flag when the half_sync_itv_timer has expired    : no
Has half_sync_itv_timer expired twice            : no
Has syncReceiptTimeoutTime expired               : no
PortAnnounceReceive state machine variables:
Current state                                    : RECEIVE
An announce message is ready to be processed     : no
PortAnnounceInformation state machine variables:
Current state                                    : POST_DISABLED
Expired announce information                     : yes
PortAnnounceTransmit state machine variables:
Current state                                    : POST_IDLE
Trigger announce information                     : no

Statistics:
Sync messages received                 : 0
Follow Up messages received            : 0
Path Delay Request messages received   : 0
Path Delay Response messages received  : 0
Path Delay messages threshold exceeded : 0
Path Delay Follow Up messages received : 0
Announce messages received             : 0
ptp messages discarded                 : 0
Sync reception timeout                 : 0
Announce reception timeout             : 0
Path Delay Requests without a response : 173
Sync messages sent                     : 0
Follow Up messages sent                : 0
Path Delay Request messages sent       : 174
Path Delay Response messages sent      : 0
Path Delay Response FUP messages sent  : 0
Announce messages sent                 : 0
```

mmahadevan108 replied:
@hakehuang could you please log an issue about this gPTP failure as it exists on the master branch as well.

@MaureenHelm Can we proceed with update to SDK 2.8. All networking tests are passing except for gptp

hakehuang replied:
@mmahadevan108  https://github.com/zephyrproject-rtos/zephyr/issues/29599 created

MaureenHelm replied:
@jukkar It appears that the gPTP failure also exists on master. Would you be ok approving this PR and handling the gPTP issue separately in #29599?

jukkar replied:
@MaureenHelm the issue with gPTP is only seen in frdm-k64f board. gPTP works ok in Atmel sam-e70 (I tested it few days ago). I do not mind you merging this, after all this is NXP board, but I will not give +1 either. The gPTP has been broken either because of some HAL update or some other earlier commit to mcux driver.

mmahadevan108 replied:
@jukkar I have gone back to release 2.3 and still see the console issue. Can you please let us know when frdm-k64f last worked for you?

jukkar replied:
@mmahadevan108 sorry, cannot recall last time I saw it working. Note that there seems to be two issues that are perhaps related to each other:
* console does not show output, although I am seeing something right now
* network interface is down even if I have network cable connected
```
** Booting Zephyr OS build zephyr-v2.4.0-1085-g007677e67645  ***


[00:00:00.003,000] <wrn> net_if: iface 0x20000f60 is down
[00:00:00.007,000] <inf> net_config: Initializing network
[00:00:00.007,000] <inf> net_config: Waiting interface 1 (0x20000f60) to be up...
[00:00:01.005,000] <wrn> net_gptp: Reset Pdelay requests
[00:00:01.008,000] <wrn> net_if: iface 0x20000f60 is down
uart:~$ 
```

The console does not respond to anything after this output, perhaps the device is hang at this point.

What I did to compile gptp sample:
```
cd samples/net/gptp/
cmake -B build -DBOARD=frdm_k64f .
make -C build flash
```

Mani-Sadhasivam reviewed (approved):
i.MX8 changes LGTM!

agansari commented:
@mmahadevan108 please rebase, #28874 has been merged

mmahadevan108 commented:
> @mmahadevan108 please rebase, #28874 has been merged

Thank you. I have rebased and fixed the merge conflict

jukkar commented:
Should gPTP `samples/net/gptp` app work with `mimxrt1050_evk` board? I assume not as this does not seem to work with upstream either. I am getting
```
zephyr/drivers/ethernet/eth_mcux.c:130:2: error: unknown type name 'enet_ptp_config_t'
  130 |  enet_ptp_config_t ptp_config;
      |  ^~~~~~~~~~~~~~~~~
```

For `frdm_k64f`, I do not get console and also do not see any network traffic. But this is also with upstream which does not seem to work properly any more. Has there been some changes recently that would affect this?

mmahadevan108 commented:
> Should gPTP `samples/net/gptp` app work with `mimxrt1050_evk` board? I assume not as this does not seem to work with upstream either. I am getting
> 
> ```
> zephyr/drivers/ethernet/eth_mcux.c:130:2: error: unknown type name 'enet_ptp_config_t'
>   130 |  enet_ptp_config_t ptp_config;
>       |  ^~~~~~~~~~~~~~~~~
> ```
> 
> For `frdm_k64f`, I do not get console and also do not see any network traffic. But this is also with upstream which does not seem to work properly any more. Has there been some changes recently that would affect this?

The build error on MXRT1050 has been fixed in the NXP HAL, the `west.yml` file needs to be updated to pull in the fix. I am awaiting a response on the console issue as I am seeing the same.

MaureenHelm commented on west.yml:
```diff
@@ -102,7 +102,7 @@ manifest:
       path: tools/net-tools
     - name: hal_nxp
-      revision: 83e5a13640417a2efb26a31f88b232e4dac31fb2
+      revision: pull/56/head
```
The hal PR has now been merged, please update this to the new SHA

mmahadevan108 replied:
Done, thank you

mmahadevan108 commented:
I have updated the PR to resolve merge conflict.
